src/ParallelAlgorithms/DiscontinuousGalerkin/FluxCommunication.hpp

```diff
@@ -0,0 +1,379 @@
+ private:
+  static constexpr size_t volume_dim = BoundaryScheme::volume_dim;
+  using temporal_id_tag = typename BoundaryScheme::temporal_id_tag;
+  using fluxes_inbox_tag = dg::FluxesInboxTag<BoundaryScheme>;
+  using all_mortar_data_tag =
+      ::Tags::Mortars<typename BoundaryScheme::mortar_data_tag, volume_dim>;
+
+ public:
+  using inbox_tags = tmpl::list<fluxes_inbox_tag>;
+
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTags>&&> apply(
+      db::DataBox<DbTags>& box, tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    if (get<domain::Tags::Element<volume_dim>>(box).number_of_neighbors() ==
+        0) {
+      return {std::move(box)};
+    }
+    auto& inbox = tuples::get<fluxes_inbox_tag>(inboxes);
+    const auto& temporal_id = get<temporal_id_tag>(box);
+    const auto temporal_received = inbox.find(temporal_id);
+    db::mutate<all_mortar_data_tag>(
+        make_not_null(&box),
+        [&temporal_received](
+            const gsl::not_null<db::item_type<all_mortar_data_tag>*>
+                mortar_data) noexcept {
+          for (auto& received_mortar_data : temporal_received->second) {
+            const auto& mortar_id = received_mortar_data.first;
+            mortar_data->at(mortar_id).remote_insert(
+                temporal_received->first,
+                std::move(received_mortar_data.second.second));
+          }
+        });
+    inbox.erase(temporal_received);
+    return {std::move(box)};
+  }
+
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex>
+  static bool is_ready(
+      const db::DataBox<DbTags>& box,
+      const tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/) noexcept {
+    return has_received_from_all_mortars<fluxes_inbox_tag>(
+        db::get<temporal_id_tag>(box),
+        get<domain::Tags::Element<volume_dim>>(box), inboxes);
+  }
+};
+
+/*!
+ * \ingroup ActionsGroup
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Receive asynchronous boundary data needed for fluxes from neighbors.
+ *
+ * Waits until remote boundary data from all internal mortars has arrived, then
+ * contributes it to the data in `Tags::Mortars<BoundaryScheme::mortar_data_tag,
+ * volume_dim>` that already holds the local boundary data.
+ *
+ * This template specialization handles asynchronous boundary schemes, e.g.
+ * for local time-stepping. It proceeds only once the received boundary data
+ * on every mortar has caught up with the `BoundaryScheme::receive_temporal_id`
+ * on that mortar.
```

Comment:
Not quite.  The received boundary data has to catch up with the BoundaryScheme::receive_temporal_id on the element, not the mortar.  The one on the mortar represents when the next message is expected from that neighbor, which is the same as the value that that neighbor has for itself.  (Although these can't be updated until the action runs, so the is_ready function has to compute the value that they will have after the inbox is processed).